labs/a1.md

```diff
@@ -0,0 +1,139 @@
+
+(regex)
+
+`grep` has options to match whole words, and to print only match-contents (instead of a whole line containing a match). You need not use "regex-grouping", just a pattern. 
+
+`sed` may also be useful for filtering by whitelisting and/or blacklisting characters. In either case, character classes "[_]" and (defining them negatively) "[^_]" can help. In both cases, use, but beware the `-` dash character: it gets interpreted as a range and triggers cryptic errors when that range is invalid. `tr`, `cut`, `sort`, and `uniq` also may be
+useful.
+
+1. What are the top 10 words, and their frequency? (for the google form)
+
+
+2. Now that you know how to extract frequency distributions, create one per line.  (Recall techniques mentioned in lecture).
+
+
+
+## niche tools & little support
+Fixing a make file. You'll get away without doing this today, but oftentimes, other people's minimally packaged code can be a good resource.
+
+These tools are available in a [packaged form][brew], but for this lab we're having you fetch them manually.
+
+\(compile some obscure text-compose too with no apt or brew formula.\)
+
+
+### jq. jshawn.
+These tools allow you to parse, restructure, and create JSON documents on the command line. Today, we're using `jq`. Find and download the utility.
+
+- [ ] Get location, lat long from a json api.
+    curl 'http://api.geonames.org/postalCodeSearchJSON?postalcode=12345&maxRows=10&username=ocf_decal' -o location.json
+
+- [ ] Parse out values with jq
+    cat location.json  \
+        | jq '.postalCodes[] | select(.placeName=="Berkeley") | {"lat": .lat, "long": .lng}'
+
+    location.json -> { key: [ {...}, {...}] }
+    jq '.key'   -> [1,2]
+    jq '.key[]' -> 1,2
+
+    jq '.[] | select(.key == "value")'
+
+    jq '. | { "k1": .key1, "k2": .key2 }'
+
+- [ ] Get ISS flyovers by location.
+Docs:
+    http://open-notify.org/Open-Notify-API/ISS-Pass-Times/ 
+
+Consider using the pipemill pattern with `date`. It can take an argument for time to display, but it requires some syntax mangling you can find in the man pages.
+
+When is the next flyovers of Berkeley? Berlin?
+
+\(Look up at the sky.\)
+
+
+#### Other APIs
+SpaceX has a [beautiful API](https://github.com/r-spacex/SpaceX-API) presenting information about at least launches.
+Something you found fun or interesting (optional)
+
+    https://github.com/toddmotto/public-apis
+    https://github.com/jdorfman/awesome-json-datasets
+    https://www.data.gov/
+
+
+### textql. or q.
+These tools allow you to query csv documents with SQL. 
+
+Let's use q. Download it from
+
+    http://harelba.github.io/q/
+
+Remember, you can run it locally as `./q` without modifying your path.
+
+    ps -ef | q -H "SELECT UID, COUNT(*) cnt FROM - GROUP BY UID ORDER BY cnt DESC LIMIT 3"
```

Comment:
there's no guarantee the advanced students actually know any SQL

Reply:
yeah, i think we should remove this section. we can't assume they know any SQL and I don't think we need them to learn it.